File: src/136_single_number/SingleNumber.java
import java.util.Arrays;

/**
 * Java skeleton for LeetCode 136 - Single Number
 *
 * Conventions:
 * - `Solution` contains the algorithm signature only (no implementation).
 * - `SingleNumber` contains `main` and test scaffolding.
 *
 * The method in `Solution` intentionally returns a default value (no implementation).
 * Tests print results in the format: actual | expected | PASS/FAIL
 */

class Solution {

    /**
     * Return the element that appears only once in the array while every other
     * element appears exactly twice.
     *
     * Signature only — no implementation provided in the skeleton.
     *
     * @param nums input array
     * @return the single number (default value provided here)
     */
    public int singleNumber(int[] nums) {
        int ans = 0;

        for (int num : nums) {
            ans ^= num;
        }

        return ans;
    }
}

public class SingleNumber {

    private static void printResult(int actual, int expected) {
        System.out.println(
            actual +
                " | " +
                expected +
                " | " +
                (actual == expected ? "PASS" : "FAIL")
        );
    }

    private static int[] cloneArray(int[] a) {
        return Arrays.copyOf(a, a.length);
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        int[][] inputs = {
            { 2, 2, 1 },
            { 4, 1, 2, 1, 2 },
            { 1 },
            { 0, 1, 0, 1, 99 },
            { -1, -1, 2 },
            { 17, 12, 5, 12, 5 },
        };

        int[] expected = { 1, 4, 1, 99, 2, 17 };

        for (int i = 0; i < inputs.length; i++) {
            int[] inp = cloneArray(inputs[i]);
            int exp = expected[i];
            int actual = sol.singleNumber(inp);
            printResult(actual, exp);
        }
    }
}

File: src/136_single_number/single_number.cpp
/leetcode/src/136_single_number/single_number.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    // Return the element that appears only once while every other element appears twice.
    // TODO: implement the algorithm (e.g., XOR scan).
    int singleNumber(vector<int>& nums) {
        int ans = 0;

        for (int num : nums) {
            ans ^= num;
        }

        return ans;
    }
};

static void print_result(int actual, int expected) {
    cout << actual << " | " << expected << " | " << (actual == expected ? "PASS" : "FAIL") << "\n";
}

int main() {
    Solution sol;

    vector<pair<vector<int>, int>> tests = {
        {{2, 2, 1}, 1},
        {{4, 1, 2, 1, 2}, 4},
        {{1}, 1},
        {{0, 1, 0}, 1},
        {{-1, -1, -2}, -2},
        {{17, 12, 5, 12, 5}, 17}
    };

    for (auto &tc : tests) {
        vector<int> nums = tc.first;
        int expected = tc.second;
        int actual = sol.singleNumber(nums);
        print_result(actual, expected);
    }

    return 0;
}
